### ISSUE TITLE
`stat` crashes on Android

### ISSUE THREAD

**Issue by Mivik** [bytecodealliance/cap-std#311]
Syscalls `open2` and `statx` are not available on some Android devices and thus calling them causes seccomp to kill the process directly.

Here are the crash logs from two devices:

```
04-09 16:31:57.811 19810 19810 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
04-09 16:31:57.811 19810 19810 F DEBUG   : Build fingerprint: 'HUAWEI/HMA-AL00/HWHMA:10/HUAWEIHMA-AL00/10.1.0.163C00:user/release-keys'
04-09 16:31:57.812 19810 19810 F DEBUG   : Revision: '0'
04-09 16:31:57.812 19810 19810 F DEBUG   : ABI: 'arm64'
04-09 16:31:57.814 19810 19810 F DEBUG   : SYSVMTYPE: Maple
04-09 16:31:57.814 19810 19810 F DEBUG   : APPVMTYPE: Art
04-09 16:31:57.815 19810 19810 F DEBUG   : Timestamp: 2023-04-09 16:31:57+0800
04-09 16:31:57.815 19810 19810 F DEBUG   : pid: 19607, tid: 19649, name: Thread-3  >>> org.flos.phira <<<
04-09 16:31:57.815 19810 19810 F DEBUG   : uid: 10397
04-09 16:31:57.815 19810 19810 F DEBUG   : signal 31 (SIGSYS), code 1 (SYS_SECCOMP), fault addr --------
04-09 16:31:57.815 19810 19810 F DEBUG   : Cause: seccomp prevented call to disallowed arm64 system call 291
04-09 16:31:57.815 19810 19810 F DEBUG   :     x0  0000000000000066  x1  0000007487e2af40  x2  0000000000000100  x3  0000000000000fff
04-09 16:31:57.815 19810 19810 F DEBUG   :     x4  0000007487e2ac60  x5  0000000074617473  x6  0000000074617473  x7  0000000078746174
04-09 16:31:57.815 19810 19810 F DEBUG   :     x8  0000000000000123  x9  0000000000000001  x10 0000000000004001  x11 0000000000000000
04-09 16:31:57.815 19810 19810 F DEBUG   :     x12 0000007580e21223  x13 0000007580e23a00  x14 00000074881b45cc  x15 0000000000000003
04-09 16:31:57.815 19810 19810 F DEBUG   :     x16 0000007488d592e8  x17 000000757e42cba0  x18 0000007486fd2000  x19 0000007487e2ae38
04-09 16:31:57.815 19810 19810 F DEBUG   :     x20 0000007487e2af40  x21 0000000000000fff  x22 0000000000000100  x23 0000000000000066
04-09 16:31:57.815 19810 19810 F DEBUG   :     x24 00000074e8fed840  x25 0000007487e2af40  x26 0000000000000025  x27 0000007488d66000
04-09 16:31:57.815 19810 19810 F DEBUG   :     x28 00000074880ce58e  x29 0000000000000000
04-09 16:31:57.815 19810 19810 F DEBUG   :     sp  0000007487e2ab60  lr  00000074889908e8  pc  000000757e42cbc0
```

```
04-09 00:34:33.772  5968  5968 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
04-09 00:34:33.772  5968  5968 F DEBUG   : Build fingerprint: 'Redmi/rubens/rubens:13/TP1A.220624.014/V14.0.23.4.2.DEV:user/release-keys'
04-09 00:34:33.772  5968  5968 F DEBUG   : Revision: '0'
04-09 00:34:33.772  5968  5968 F DEBUG   : ABI: 'arm64'
04-09 00:34:33.772  5968  5968 F DEBUG   : Timestamp: 2023-04-09 00:34:32.513516823+0800
04-09 00:34:33.772  5968  5968 F DEBUG   : Process uptime: 6s
04-09 00:34:33.772  5968  5968 F DEBUG   : Cmdline: org.phira
04-09 00:34:33.772  5968  5968 F DEBUG   : pid: 5723, tid: 5755, name: Thread-4  >>> org.phira <<<
04-09 00:34:33.772  5968  5968 F DEBUG   : uid: 10195
04-09 00:34:33.772  5968  5968 F DEBUG   : tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
04-09 00:34:33.772  5968  5968 F DEBUG   : signal 31 (SIGSYS), code 1 (SYS_SECCOMP), fault addr --------
04-09 00:34:33.772  5968  5968 F DEBUG   : Cause: seccomp prevented call to disallowed arm64 system call 437
04-09 00:34:33.772  5968  5968 F DEBUG   :     x0  0000000000000094  x1  00000078e7d29418  x2  00000078e7d293a0  x3  0000000000000018
04-09 00:34:33.772  5968  5968 F DEBUG   :     x4  000000000000000a  x5  00000000666e693a  x6  00000000666e693a  x7  000000006f666e69
04-09 00:34:33.772  5968  5968 F DEBUG   :     x8  00000000000001b5  x9  0000000000000000  x10 0000000a00000000  x11 000000000000000a
04-09 00:34:33.772  5968  5968 F DEBUG   :     x12 0000000000000000  x13 00000000ffffffbf  x14 00000078ea360b99  x15 0000000000000003
04-09 00:34:33.772  5968  5968 F DEBUG   :     x16 00000078eae163b8  x17 0000007a046a40c0  x18 00000078e78e8000  x19 00000078e7d29630
04-09 00:34:33.772  5968  5968 F DEBUG   :     x20 b4000078db6322d0  x21 0000000000080000  x22 0000000000000000  x23 0000000000000005
04-09 00:34:33.772  5968  5968 F DEBUG   :     x24 00000078e7d29418  x25 0000000000000006  x26 0000000000000004  x27 0000000500000000
04-09 00:34:33.772  5968  5968 F DEBUG   :     x28 00000078eae23000  x29 0000000000000000
04-09 00:34:33.772  5968  5968 F DEBUG   :     lr  00000078eaa50610  sp  00000078e7d293a0  pc  0000007a046a40e0  pst 0000000000001000
```

`cap_std` version: `1.0.12`

**Comment by sunfishcode** [comment]
Thanks for the report! cap-std already has a version check for `openat2`, but didn't have one yet for `statx`. I've now submitted #312 to fix that.

**Comment by Hoodad** [comment]
> Thanks for the report! cap-std already has a version check for `openat2`, but didn't have one yet for `statx`. I've now submitted #312 to fix that.

I'm not sure if that is the correct way of checking if certain sys calls is supported as some are just not allowed by Android. See [list](https://cs.android.com/android/platform/superproject/+/master:bionic/libc/SYSCALLS.TXT;l=1?q=SYSCALLS.TXT) 

I was just about to write an issue similar to this about `openat2` not being allowed on certain Android devices with a similar crash as this issue. See logs below 

Please note that on some Android devices the code runs just fine, but on newer device it crashes hard. This particular log is from an Samsung S23 where as the same code on a Samsung S20 works just fine. The crash has also been noticed on Pixel 6 device.

I think what's needed is some verification with what is allowed by the [Android Seccomp](https://android-developers.googleblog.com/2017/07/seccomp-filter-in-android-o.html) 

```text
04-04 14:18:50.859 32620 32620 F DEBUG   : Build fingerprint: 'samsung/dm1qxeea/dm1q:13/TP1A.220624.014/S911BXXU1AWC8:user/release-keys'
04-04 14:18:50.859 32620 32620 F DEBUG   : Revision: '13'
04-04 14:18:50.859 32620 32620 F DEBUG   : ABI: 'arm64'
04-04 14:18:50.859 32620 32620 F DEBUG   : Processor: '6'
04-04 14:18:50.859 32620 32620 F DEBUG   : Timestamp: 2023-04-04 14:18:50.705002336+0200
04-04 14:18:50.859 32620 32620 F DEBUG   : Process uptime: 1s
04-04 14:18:50.859 32620 32620 F DEBUG   : Cmdline: rust.XXXXXX
04-04 14:18:50.859 32620 32620 F DEBUG   : pid: 32572, tid: 32602, name: Thread-3  >>> rust.XXXXXX<<<
04-04 14:18:50.859 32620 32620 F DEBUG   : uid: 10510
04-04 14:18:50.859 32620 32620 F DEBUG   : tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
04-04 14:18:50.859 32620 32620 F DEBUG   : pac_enabled_keys: 000000000000000f (PR_PAC_APIAKEY, PR_PAC_APIBKEY, PR_PAC_APDAKEY, PR_PAC_APDBKEY)
04-04 14:18:50.859 32620 32620 F DEBUG   : signal 31 (SIGSYS), code 1 (SYS_SECCOMP), fault addr --------
04-04 14:18:50.859 32620 32620 F DEBUG   : Cause: seccomp prevented call to disallowed arm64 system call 437
04-04 14:18:50.859 32620 32620 F DEBUG   :     x0  000000000000006f  x1  0000007ba6df61f0  x2  0000007ba6df5ff0  x3  0000000000000018
04-04 14:18:50.859 32620 32620 F DEBUG   :     x4  000000000000000a  x5  0000007ba6df61f6  x6  0000007ba6df61f6  x7  7f7f7f7f7f7f7f7f
04-04 14:18:50.859 32620 32620 F DEBUG   :     x8  00000000000001b5  x9  0000000000000000  x10 0000000f00000000  x11 000000000000000a
04-04 14:18:50.859 32620 32620 F DEBUG   :     x12 0000000000000005  x13 00000000ffffffbf  x14 0000007ba7bf0407  x15 0000000000000003
04-04 14:18:50.859 32620 32620 F DEBUG   :     x16 0000007baaf77af0  x17 0000007cbe2ef610  x18 0000007ba3ecc000  x19 0000007ba6df63f8
04-04 14:18:50.859 32620 32620 F DEBUG   :     x20 0000007ba6df6eb8  x21 0000000000284000  x22 0000000000000000  x23 0000000000000007
04-04 14:18:50.859 32620 32620 F DEBUG   :     x24 0000007ba6df61f0  x25 0000000000000008  x26 0000000000000004  x27 0000000500000000
04-04 14:18:50.859 32620 32620 F DEBUG   :     x28 0000007baafc6000  x29 0000000000000000
04-04 14:18:50.859 32620 32620 F DEBUG   :     lr  0000007baa9e3b04  sp  0000007ba6df5ff0  pc  0000007cbe2ef634  pst 0000000000001000
04-04 14:18:50.859 32620 32620 F DEBUG   : backtrace:
04-04 14:18:50.859 32620 32620 F DEBUG   :       #00 pc 0000000000089634  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+36) (BuildId: 1bcad8bca80d38bceb9089f70d394e33)
04-04 14:18:50.859 32620 32620 F DEBUG   :       #01 pc 00000000039e1b00  /data/app/~~qwCVulGe1iQ4hdviwPPobA==/rust.XXXXXX-tJ_t0m_miM1rVLX7dFvrbw==/lib/arm64/XXXXXX.so (rustix::backend::fs::syscalls::openat2::h1eae23f9f1aad1f3+44) (BuildId: 5f2d67357670054fe6c7f2de0335d56429e9f63b)
```

**Comment by sunfishcode** [comment]
Thanks for pointing that out. From that documentation, there doesn't seem to be a way to detect this without crashing the process. I've now updated https://github.com/bytecodealliance/cap-std/pull/312 to disable the use of `statx` and `openat2` on Android altogether.

**Comment by Hoodad** [comment]
Hmm that's sad maybe they won't you to parse the SYSCALLS.txt file if that is possible? I'm speculating...

I tried in a fork to report `opencat2` to never be supported [here](https://github.com/EmbarkStudios/cap-std/tree/dont-use-opencat2-on-android) and that seems to work fine without any immediate issues so I think your change to straight out disable the sys call on Android is a suitable solution👍 

Looking forward to next release with this included 😃

**Comment by sunfishcode** [comment]
I didn't see a mention of SYSCALLS.txt being installed on devices in a place we could read it. But if anyone has any ideas on what we could do here, I'm open to them.

**Comment by sunfishcode** [comment]
This is now released in 1.0.13.

**Comment by eddyb** [comment]
This needs to be reopened, as `stat_impl` calling into `open_beneath` was overlooked.
(I left more detailed comments at https://github.com/bytecodealliance/cap-std/pull/312#pullrequestreview-1387770855, on the PR that landed)

**Comment by sunfishcode** [comment]
And the other uses of `openat2` are now fixed too, and released in 1.0.14.
